From 3718655a402a2d3fce84a6b3223a00124b8518e7 Mon Sep 17 00:00:00 2001
From: Mark Van Holstyn <mvanholstyn@gmail.com>
Date: Fri, 23 Jan 2009 16:05:54 -0500
Subject: [PATCH] Update gems:build to only build gems configured for the current Rails.env.

This will allow configuration of gems in the test environment without requiring the necessary libraries to be installed on the production server to build those gems.
---
 railties/lib/tasks/gems.rake |   11 ++++++-----
 1 files changed, 6 insertions(+), 5 deletions(-)

diff --git a/railties/lib/tasks/gems.rake b/railties/lib/tasks/gems.rake
index 754e3ba..876d01f 100644
--- a/railties/lib/tasks/gems.rake
+++ b/railties/lib/tasks/gems.rake
@@ -22,14 +22,15 @@ namespace :gems do
   end
 
   desc "Build any native extensions for unpacked gems"
-  task :build do
-    $rails_gem_installer = true
+  task :build => :base do
     require 'rails/gem_builder'
-    Dir[File.join(Rails::GemDependency.unpacked_path, '*')].each do |gem_dir|
-      spec_file = File.join(gem_dir, '.specification')
+    gems_to_build = Rails.configuration.gems + Rails.configuration.gems.map { |gem| gem.dependencies }.flatten
+    gems_to_build.uniq.each do |gem|
+      next unless gem.frozen? && (ENV['GEM'].blank? || ENV['GEM'] == gem.name)
+      gem_dir = gem.gem_dir(Rails::GemDependency.unpacked_path)
+      spec_file = gem.spec_filename(Rails::GemDependency.unpacked_path)
       next unless File.exists?(spec_file)
       specification = YAML::load_file(spec_file)
-      next unless ENV['GEM'].blank? || ENV['GEM'] == specification.name
       Rails::GemBuilder.new(specification, gem_dir).build_extensions
       puts "Built gem: '#{gem_dir}'"
     end
-- 
1.6.1

